File: Day189_2025/Q189_Leetcode.py
from typing import List


class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        events.sort()
        n = len(events)
        dp = [[0] * (n + 1) for _ in range(k + 1)]

        def binary_search(target):
            left, right = 0, n
            while left < right:
                mid = (left + right) // 2
                if events[mid][0] <= target:
                    left = mid + 1
                else:
                    right = mid
            return left

        for i in range(n - 1, -1, -1):
            for cnt in range(1, k + 1):
                nxt = binary_search(events[i][1])
                dp[cnt][i] = max(dp[cnt][i + 1], events[i][2] + dp[cnt - 1][nxt])
        return dp[k][0]

File: Day189_2025/Q189_Leetcode.java
import java.util.*;

class Q189_Leetcode {
    public int maxValue(int[][] events, int k) {
        int n = events.length;
        Arrays.sort(events, Comparator.comparingInt(a -> a[0]));

        int[][] dp = new int[k + 1][n + 1];

        for (int cur = n - 1; cur >= 0; cur--) {
            for (int cnt = 1; cnt <= k; cnt++) {
                int next = binarySearch(events, events[cur][1]);
                dp[cnt][cur] = Math.max(dp[cnt][cur + 1], events[cur][2] + dp[cnt - 1][next]);
            }
        }
        return dp[k][0];
    }

    private int binarySearch(int[][] events, int target) {
        int left = 0, right = events.length;
        while (left < right) {
            int mid = (left + right) / 2;
            if (events[mid][0] <= target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
